chore(eslint): enable typescript 3.7 rules

JamieMagee opened the pull request:
Enable `prefer-optional-chain` and `prefer-nullish-coalescing` and fix and occurences.

No failures with `prefer-nullish-coalescing` found. Might need to look at the options for that rule.

viceice commented on lib/platform/gitlab/index.ts:
```diff
@@ -479,8 +479,9 @@ export async function getPr(iid: number): Promise<Pr> {
     ).body;
-    const branchCommitEmail =
-      branch && branch.commit ? branch.commit.author_email : null;
+    const branchCommitEmail = branch?.commit
+      ? branch.commit.author_email
+      : null;
```
```suggestion
    const branchCommitEmail = branch?.commit?.author_email ?? null;
```

🤔 I think `branch` can't be `undefined` here. The `api.getJson` call would fail then.

rarkins commented on lib/config/migration.ts:
```diff
@@ -336,12 +336,7 @@ export function migrateConfig(
-              parsedSchedule.t_b &&
-              parsedSchedule.t_a[0] > parsedSchedule.t_b[0]
-            ) {
+            if (parsedSchedule?.t_a[0] > parsedSchedule?.t_b[0]) {
```
Needs ?. before [

viceice replied:
```suggestion
            if (parsedSchedule?.t_a?.[0] > parsedSchedule?.t_b?.[0]) {
```
sure

rarkins commented on lib/datasource/npm/get.ts:
```diff
@@ -88,10 +88,10 @@
~    regUrl.replace(/\/?$/, '/') === npmrc.registry?.replace(/\/?$/, '/')
~  ) {
     authInfo = { type: 'Bearer', token: npmrc._authToken };
   }
```
Second ?. is unnecessary

viceice replied:
sure

viceice commented on lib/datasource/npm/get.ts:
```diff
@@ -91,7 +91,7 @@ export async function getDependency(
   }
 
-  if (authInfo && authInfo.type && authInfo.token) {
+  if (authInfo?.type && authInfo?.token) {
```
```suggestion
  if (authInfo?.type && authInfo.token) {
```

JamieMagee commented:
I think I've addressed all your comments. Also there's a major version bump for eslint waiting, which might solve some of the issues with the nullish coalescing rule.

rarkins commented:
BTW I was looking for this eslint rule recently and didn't find it :)

Does it "auto fix" ?

JamieMagee commented:
> some snapshots are failed

I think this might be an error with the most recent commit. I'll check locally.

> BTW I was looking for this eslint rule recently and didn't find it :)
> 
> Does it "auto fix" ?

You asking prompted me to look :smile: 

It does have an auto-fix, but it's still quite buggy and only recommended for the simplest cases. I did all of this by hand.

viceice commented:
> I think I've addressed all your comments. Also there's a major version bump for eslint waiting, which might solve some of the issues with the nullish coalescing rule.

Yes, we should update to new major linter versions, `@types/eslint` should be grouped to linters before. I'll see if i can do that at weekend. Or @JamieMagee have you time to arrange that?

viceice commented on lib/platform/gitlab/index.ts:
```diff
@@ -479,7 +479,7 @@ export async function getPr(iid: number): Promise<Pr> {
       await gitlabApi.getJson<{ commit: { author_email: string } }>(branchUrl)
     ).body;
-    const branchCommitEmail = branch.commit.author_email;
+    const branchCommitEmail = branch?.commit?.author_email ?? null;
```
Looks like we have a bad test? We can ignore that for now because #6697 will remove this anyways

JamieMagee replied:
Yeah, it's been like that for so long, I can live with it until we merge that PR.
